### PR TITLE
Support dynamic shapes for IndexedSlices in the _num_elements function

### LINKED ISSUE #33681: _num_elements(grad) difference between Tensor and IndexedSlices behavior
<em>Please make sure that this is a bug. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Linux-4.4.0-87-generic-x86_64-with-Ubuntu-16.04-xenial
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: N/A
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): 2.0
- Python version: 3.7
- Bazel version (if compiling from source): N/A
- GCC/Compiler version (if compiling from source): N/A
- CUDA/cuDNN version: N/A
- GPU model and memory: N/A

You can collect some of this information using our environment capture
[script](https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh)
You can also obtain the TensorFlow version with: 1. TF 1.0: `python -c "import
tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"` 2. TF 2.0: `python -c
"import tensorflow as tf; print(tf.version.GIT_VERSION, tf.version.VERSION)"`

**Describe the current behavior**
When training a model (which I can't make public at the moment), I've got an error: `<built-in function len> returned a result with an error set`. Debugging showed the reason was in
https://github.com/tensorflow/tensorflow/blob/0d76b0f988d3d7b06f16da950b3ed95e4ec59e82/tensorflow/python/eager/backprop.py#L615-L624
Namely, my code ended up with `grad` being an instance of `IndexedSlices`, but `grad.values._shape_tuple()` was `(None, 16)`. 

Changing code to
```
    shape_tuple = grad.values._shape_tuple()  # pylint: disable=protected-access
    if shape_tuple is None or None in shape_tuple:
      return 0
    return functools.reduce(operator.mul, shape_tuple, 1)
```
by analogy with the `Tensor` case appears to work and let my model train, but I am not certain whether that's correct.

**Describe the expected behavior**
Either handle the case where `grad.values._shape_tuple()` is/contains `None`, or explicitly throw `ValueError` if it shouldn't be allowed.

**Code to reproduce the issue**
(Will try to do this later)

**Other info / logs**
Include any logs or source code that would be helpful to diagnose the problem. If including tracebacks, please include the full traceback. Large logs and files should be attached.

### PR DESCRIPTION
**Pull request by alexeyr**
Fixes #33681.

### CODE REVIEW

**Comment by rthadur** [comment]
@alexeyr is it possible to write some test cases around these changes ?

**Comment by alexeyr** [comment]
@rthadur Can you say where should I look for the current tests of this functionality?

**Comment by alextp** [comment]
@alexeyr tensorflow/python/eager/backprop_test.py contains the tests for backprop.py

**Review by jaingaurav — DISMISSED** [review]

**Comment by rthadur** [comment]
@alexeyr can you please fix the build failures ?

**Comment by alexeyr** [comment]
Out of work for family reasons, I'll look into it tomorrow.

**Comment by alexeyr** [comment]
@rthadur Well, that... sure was an error, should be fixed now. Still trying to write a proper test.

**Comment by rthadur** [comment]
@alexeyr thank you , should i mark this as WIP till test cases are done ?

**Comment by alexeyr** [comment]
Oh, #34093 can provide a test. I'll try it on Monday, if not earlier.

**Comment by jaingaurav** [comment]
@alexeyr: Yes, but I think ideally we'll want something a bit more isolated to this change if possible.

**Comment by alexeyr** [comment]
Ok, so I need to somehow make sure `InitialGradients` 
https://github.com/tensorflow/tensorflow/blob/0250bfdb8c8f0dce3e11c91f5969c994660d21d9/tensorflow/c/eager/tape.h#L566-L573
fills an entry in its `result` with more than `kMinAggregateCount` (i.e. 4) values. Can't figure out how to make it happen yet :(

**Comment by gbaned** [comment]
@jaingaurav Can you please take a look on this PR? Thanks!

**Comment by gbaned** [comment]
@alexeyr Can you please resolve conflicts? Thanks!

**Comment by tensorflowbutler** [comment]
It has been 15 days with no activity and the `awaiting response` label was assigned. Is this PR still valid? Assigning the `stalled` label. Please comment to reassure me that this is still being worked on.

**Comment by gbaned** [comment]
I'm going to go ahead and close this PR, because it seems to have stalled. If you're still interested in pursing this (and responding to my comments), please feel free to reopen!
